<h1>NLP with Support Vecter Machine And Naive Bays </h1>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
from sklearn.datasets import fetch_20newsgroups
twenty_train = fetch_20newsgroups(subset='train',shuffle=True)


In [3]:
from pprint import pprint
pprint(list(twenty_train.target_names))

['alt.atheism',
 'comp.graphics',
 'comp.os.ms-windows.misc',
 'comp.sys.ibm.pc.hardware',
 'comp.sys.mac.hardware',
 'comp.windows.x',
 'misc.forsale',
 'rec.autos',
 'rec.motorcycles',
 'rec.sport.baseball',
 'rec.sport.hockey',
 'sci.crypt',
 'sci.electronics',
 'sci.med',
 'sci.space',
 'soc.religion.christian',
 'talk.politics.guns',
 'talk.politics.mideast',
 'talk.politics.misc',
 'talk.religion.misc']


In [4]:
group_names = ['atheism','graphics','MS windows','pc hardware','mac hardware','windows','forsale','autos','motorcycles','baseball',
              'hockey','cryptography','electronics','medical','space','christian','guns','mideast','politicsmisc','religen msc']

<style>
    p {
        
        color:'blue';
            
        
    }

</style>

In [107]:
# Printing all the Catogaries
twenty_train.target_names
print((twenty_train.data[48][:]))

From: tg@cs.toronto.edu (Tom Glinos)
Subject: 12V to 3V and 48V at 3A
Organization: Department of Computer Science, University of Toronto
Distribution: na
Lines: 11

The subject line says it all. I'm working on a project
that will use a car battery. I need to pull off 3V and possibly
48V at 3A.

I have several ideas, but I'd prefer to benefit from all you
brilliant people :-)
-- 
"Conquest is easy, control is not"	| Tom Glinos @ U of Toronto Statistics
[Star Trek TOS]   			| tg@utstat.toronto.edu
USL forgot this simple history lesson



<h2>Extracting the Features From The Text</h2>

<p>Using Bag Of Words Model. Segment Each File to Words and Count number of times Each word Occors and Assign a id<p>

In [32]:
from sklearn.feature_extraction.text import CountVectorizer
count_vect = CountVectorizer()
X_train_counts = count_vect.fit_transform(twenty_train.data)
X_train_counts.shape

(11314, 130107)

In [33]:
from sklearn.feature_extraction.text import TfidfTransformer
tfid_transformer = TfidfTransformer()
X_train_tfid = tfid_transformer.fit_transform(X_train_counts)
X_train_tfid.shape

(11314, 130107)

<h2>Running The Machine Learning ALgorithems </h2>

<h3>Naive Bays</h3>

<p>Using The Naive Bays Classifier in Sklearn and tranning it </p>

In [34]:
from sklearn.naive_bayes import MultinomialNB

<h3>Accuracy of the Model With Stop Words</h3>

In [35]:
from sklearn.pipeline import Pipeline
text_clf = Pipeline([('vect', CountVectorizer()),
                    ('tfidf', TfidfTransformer()),
                    ('clf', MultinomialNB()) ])

text_clf = text_clf.fit(twenty_train.data,twenty_train.target)

In [36]:
twenty_test = fetch_20newsgroups(subset='test',shuffle=True)
predicted = text_clf.predict(twenty_test.data)

In [37]:
np.mean(predicted == twenty_test.target)

0.7738980350504514

In [12]:
from sklearn.metrics import classification_report
print(classification_report(twenty_test.target, predicted,target_names =group_names))

              precision    recall  f1-score   support

     atheism       0.80      0.52      0.63       319
    graphics       0.81      0.65      0.72       389
  MS windows       0.82      0.65      0.73       394
 pc hardware       0.67      0.78      0.72       392
mac hardware       0.86      0.77      0.81       385
     windows       0.89      0.75      0.82       395
     forsale       0.93      0.69      0.80       390
       autos       0.85      0.92      0.88       396
 motorcycles       0.94      0.93      0.93       398
    baseball       0.92      0.90      0.91       397
      hockey       0.89      0.97      0.93       399
cryptography       0.59      0.97      0.74       396
 electronics       0.84      0.60      0.70       393
     medical       0.92      0.74      0.82       396
       space       0.84      0.89      0.87       394
   christian       0.44      0.98      0.61       398
        guns       0.64      0.94      0.76       364
     mideast       0.93    

<h3>Accuracy of the Model With Out Stop Words</h3>

In [38]:
from sklearn.pipeline import Pipeline
text_clf = Pipeline([('vect', CountVectorizer(stop_words='english')),
                    ('tfidf', TfidfTransformer()),
                    ('clf', MultinomialNB()) ])

text_clf = text_clf.fit(twenty_train.data,twenty_train.target)

In [41]:

twenty_test = fetch_20newsgroups(subset='test',shuffle=True)
predicted = text_clf.predict(twenty_test.data)

In [42]:
np.mean(predicted == twenty_test.target)

0.8169144981412639

In [44]:
from sklearn.metrics import classification_report
print(classification_report(twenty_test.target, predicted,target_names =group_names))

              precision    recall  f1-score   support

     atheism       0.80      0.69      0.74       319
    graphics       0.78      0.72      0.75       389
  MS windows       0.79      0.72      0.75       394
 pc hardware       0.68      0.81      0.74       392
mac hardware       0.86      0.81      0.84       385
     windows       0.87      0.78      0.82       395
     forsale       0.87      0.80      0.83       390
       autos       0.88      0.91      0.90       396
 motorcycles       0.93      0.96      0.95       398
    baseball       0.91      0.92      0.92       397
      hockey       0.88      0.98      0.93       399
cryptography       0.75      0.96      0.84       396
 electronics       0.84      0.65      0.74       393
     medical       0.92      0.79      0.85       396
       space       0.82      0.94      0.88       394
   christian       0.62      0.96      0.76       398
        guns       0.66      0.95      0.78       364
     mideast       0.95    

<h4>In Naive Bays Removing Stop Words Incresed Accuracy from 77.38% to 81.69%</h4>

<h3>Support Vecter Machine</h3>

In [45]:
# With Stop Words 
from sklearn.linear_model import SGDClassifier

text_clf_svm = Pipeline([('vect', CountVectorizer()),
                    ('tfidf', TfidfTransformer()),
                    ('clf-svm', SGDClassifier(loss='hinge',penalty='l2',alpha=1e-3,n_iter=5,random_state=0)) ])
text_clf_svm.fit(twenty_train.data,twenty_train.target)

C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:152: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


Pipeline(memory=None,
     steps=[('vect', CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip...ndom_state=0, shuffle=True, tol=None,
       validation_fraction=0.1, verbose=0, warm_start=False))])

In [46]:

predicted_svm = text_clf_svm.predict(twenty_test.data)

In [47]:
np.mean(predicted_svm == twenty_test.target)

0.8244822092405736

In [48]:
print(classification_report(twenty_test.target, predicted_svm,target_names =group_names))

              precision    recall  f1-score   support

     atheism       0.73      0.71      0.72       319
    graphics       0.80      0.70      0.75       389
  MS windows       0.72      0.78      0.75       394
 pc hardware       0.73      0.68      0.71       392
mac hardware       0.82      0.82      0.82       385
     windows       0.84      0.76      0.80       395
     forsale       0.84      0.90      0.87       390
       autos       0.92      0.89      0.91       396
 motorcycles       0.94      0.96      0.95       398
    baseball       0.90      0.90      0.90       397
      hockey       0.88      0.99      0.93       399
cryptography       0.83      0.96      0.89       396
 electronics       0.81      0.64      0.72       393
     medical       0.88      0.85      0.86       396
       space       0.85      0.96      0.90       394
   christian       0.74      0.94      0.83       398
        guns       0.70      0.92      0.80       364
     mideast       0.90    

<h3>Removing Stop Words</h3>
<p> removing words like the, then from the data becase these are not useful for the clsssification</p>

In [49]:
# Without Stop Words
from sklearn.linear_model import SGDClassifier

text_clf_svm = Pipeline([('vect', CountVectorizer(stop_words='english')),
                    ('tfidf', TfidfTransformer()),
                    ('clf-svm', SGDClassifier(loss='hinge',penalty='l2',alpha=1e-3,n_iter=5,random_state=0)) ])
text_clf_svm.fit(twenty_train.data,twenty_train.target)

C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:152: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


Pipeline(memory=None,
     steps=[('vect', CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words='english',
        ...ndom_state=0, shuffle=True, tol=None,
       validation_fraction=0.1, verbose=0, warm_start=False))])

In [50]:
predicted_svm = text_clf_svm.predict(twenty_test.data)

In [51]:
np.mean(predicted_svm == twenty_test.target)

0.8228890069038768

In [52]:
print(classification_report(twenty_test.target, predicted_svm,target_names =group_names))

              precision    recall  f1-score   support

     atheism       0.72      0.71      0.71       319
    graphics       0.78      0.70      0.74       389
  MS windows       0.73      0.79      0.76       394
 pc hardware       0.72      0.68      0.70       392
mac hardware       0.82      0.82      0.82       385
     windows       0.86      0.76      0.80       395
     forsale       0.82      0.87      0.84       390
       autos       0.91      0.89      0.90       396
 motorcycles       0.94      0.97      0.95       398
    baseball       0.89      0.92      0.90       397
      hockey       0.87      0.98      0.93       399
cryptography       0.85      0.96      0.90       396
 electronics       0.81      0.62      0.70       393
     medical       0.90      0.87      0.88       396
       space       0.84      0.96      0.90       394
   christian       0.74      0.93      0.82       398
        guns       0.70      0.93      0.80       364
     mideast       0.92    

<h2>Optimizing the Performance Of the Model</h2>

<h3>Grid Search For Naive Bays</h3>

In [74]:
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
parameters = {'vect__ngram_range': [(1, 1), (1, 2)],
              'tfidf__use_idf': (True, False),
              'clf__alpha': [0.009,0.008,0.01,0.02]}

text_clf = Pipeline([('vect', CountVectorizer(stop_words='english')),
                    ('tfidf', TfidfTransformer()),
                    ('clf', MultinomialNB()) ])



In [75]:

gs_clf = GridSearchCV(text_clf, parameters, n_jobs=-1)
gs_clf = gs_clf.fit(twenty_train.data, twenty_train.target)

C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)


In [76]:
gs_clf.best_params_

{'clf__alpha': 0.02, 'tfidf__use_idf': True, 'vect__ngram_range': (1, 1)}

In [77]:
gs_clf.best_score_

0.905957221141948

<p>IN Naive Bays The Accuracy Of the Model Increased <br> It Was Computationlally intensive Rather than Just Using the Model <br>
    <ul>
        <li>Accuracy Normal : 77.3898 %</li>
        <li>Accuracy (After Removing Stop Words): 81.6914 % </li>
        <li>Accuracy After Applying Grid Serch: 90.5957 % </li>
        
   </ul>



</p>

<h3>Grid Search For SVM</h3>

In [85]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import SGDClassifier
parameters_svm = {'vect__ngram_range': [(1, 1), (1, 2)],
                   'tfidf__use_idf': (True, False),
                   'clf-svm__alpha': [0.0008,0.0009,0.001,0.002]}


text_clf_svm = Pipeline([('vect', CountVectorizer(stop_words='english')),
                    ('tfidf', TfidfTransformer()),
                    ('clf-svm', SGDClassifier(loss='hinge',penalty='l2',alpha=1e-3,n_iter=5,random_state=0)) ])




In [86]:

gs_clf_svm = GridSearchCV(text_clf_svm, parameters_svm, n_jobs=-1)
gs_clf_svm = gs_clf_svm.fit(twenty_train.data, twenty_train.target)


C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\stochastic_gradient.py:152: DeprecationWarning: n_iter parameter is deprecated in 0.19 and will be removed in 0.21. Use max_iter and tol instead.
  DeprecationWarning)


In [87]:
gs_clf_svm.best_params_

{'clf-svm__alpha': 0.0009, 'tfidf__use_idf': True, 'vect__ngram_range': (1, 2)}

In [88]:
gs_clf_svm.best_score_

0.8962347534028637

<p>In SVM The Accuracy Of the Model Increased only sightly compaied to Naive Bays<br> 
Training the Model with Grid Search Was Computationlally intensive Rather than Just Using the Model just <br>
    <ul>
        <li>Accuracy Normal : 82.4482 %</li>
        <li>Accuracy (After Removing Stop Words): 82.288 % </li>
        <li>Accuracy After Applying Grid Serch: 89.5881 % </li>
        
   </ul>



</p>

<h2>Logistic Regression</h2>

In [89]:
from sklearn.linear_model import LogisticRegression

logreg = Pipeline([('vect', CountVectorizer()),
                ('tfidf', TfidfTransformer()),
                ('clf', LogisticRegression(n_jobs=1, C=1e5)),
               ])
logreg.fit(twenty_train.data, twenty_train.target)



C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


Pipeline(memory=None,
     steps=[('vect', CountVectorizer(analyzer='word', binary=False, decode_error='strict',
        dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
        lowercase=True, max_df=1.0, max_features=None, min_df=1,
        ngram_range=(1, 1), preprocessor=None, stop_words=None,
        strip...penalty='l2', random_state=None,
          solver='warn', tol=0.0001, verbose=0, warm_start=False))])

In [90]:

predicted_lg = logreg.predict(twenty_test.data)

In [91]:
np.mean(predicted_lg == twenty_test.target)

0.8481147105682422

In [92]:
print(classification_report(twenty_test.target, predicted_lg,target_names =group_names))

              precision    recall  f1-score   support

     atheism       0.81      0.80      0.81       319
    graphics       0.75      0.80      0.78       389
  MS windows       0.77      0.72      0.74       394
 pc hardware       0.70      0.73      0.72       392
mac hardware       0.83      0.85      0.84       385
     windows       0.85      0.76      0.80       395
     forsale       0.82      0.89      0.85       390
       autos       0.92      0.91      0.91       396
 motorcycles       0.96      0.96      0.96       398
    baseball       0.93      0.94      0.93       397
      hockey       0.95      0.98      0.97       399
cryptography       0.94      0.93      0.94       396
 electronics       0.78      0.79      0.79       393
     medical       0.90      0.86      0.88       396
       space       0.90      0.93      0.91       394
   christian       0.85      0.93      0.89       398
        guns       0.74      0.91      0.82       364
     mideast       0.97    

<h3>Without Stop Words</h3>

In [93]:
text_clf_lg = Pipeline([('vect', CountVectorizer(stop_words='english')),
                ('tfidf', TfidfTransformer()),
                ('clf', LogisticRegression(n_jobs=1, C=1e5)),
               ])
text_clf_lg.fit(twenty_train.data, twenty_train.target)

predicted_lg = text_clf_lg.predict(twenty_test.data)
np.mean(predicted_lg == twenty_test.target)


0.8479819437068508

<h3>Using Grid Search</h3>

In [95]:
parameters = {'vect__ngram_range': [(1, 1), (1, 2)],
              'tfidf__use_idf': (True, False),
             'clf__C':[0.8,0.9,1,2,10],
             'clf__penalty': ["l1","l2"]}
              

text_clf_lg = Pipeline([('vect', CountVectorizer(stop_words='english')),
                ('tfidf', TfidfTransformer()),
                ('clf', LogisticRegression(n_jobs=1)),
                   ])

In [96]:

gs_clf_log = GridSearchCV(text_clf_lg, parameters, n_jobs=-1)
gs_clf_log = gs_clf_log.fit(twenty_train.data, twenty_train.target)

C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\model_selection\_split.py:2053: FutureWarning: You should specify a value for 'cv' instead of relying on the default value. The default value will change from 3 to 5 in version 0.22.
  warnings.warn(CV_WARNING, FutureWarning)
C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


In [97]:
gs_clf_log.best_params_

{'clf__C': 10,
 'clf__penalty': 'l2',
 'tfidf__use_idf': True,
 'vect__ngram_range': (1, 2)}

In [98]:
gs_clf_log.best_score_

0.9112603853632667

<P> In Logistic Regression again we have no notable difference while removing the Stop Words <br>
but just training the model gives us < 80% accuracy 
<ul>
        <li>Accuracy Normal : 84.8114 %</li>
        <li>Accuracy (After Removing Stop Words): 84.7981 % </li>
        <li>Accuracy After Applying Grid Serch: 91.7977 % </li>
        
</ul>

</p>

In [99]:
from sklearn.externals import joblib
joblib.dump(gs_clf_log,"Logistic_grid.sav")


['Logistic_grid.sav']

In [101]:
text_clf_logistic = Pipeline([('vect', CountVectorizer(stop_words='english',ngram_range=(1, 2))),
                ('tfidf', TfidfTransformer(use_idf= True)),
                ('clf', LogisticRegression(n_jobs=1,C=10),penalty=l2'),
                   ])

In [103]:
text_clf_logistic = text_clf_logistic.fit(twenty_train.data, twenty_train.target)

C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\smvdi\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:460: FutureWarning: Default multi_class will be changed to 'auto' in 0.22. Specify the multi_class option to silence this warning.
  "this warning.", FutureWarning)


In [105]:
log_opt = text_clf_logistic.predict(twenty_test.data)

In [106]:
np.mean(log_opt == twenty_test.target)

0.8462559745087627

In [108]:
from sklearn.externals import joblib
joblib.dump(text_clf_logistic,"Logistic_20News.sav")


['Logistic_20News.sav']